File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.py
class Solution(object):
    def minFallingPathSum(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: int
        """
        n=len(matrix)
        

        if n==1:
            return matrix[0][0] 

        for i in xrange(1,n):
            for j in xrange(n):
                ld=u=rd=1e7
                if j>=1:
                    ld=matrix[i][j]+matrix[i-1][j-1]
                u=matrix[i][j]+matrix[i-1][j]
                if j<n-1:
                    rd=matrix[i][j]+matrix[i-1][j+1]
                matrix[i][j]=min(ld,u,rd)
                
        return min(matrix[n-1])

File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.java
class Solution {
    public int minFallingPathSum(int[][] matrix) {
        int n = matrix.length;
        
        for (int i = n - 2; i >= 0; i--){
            for (int j = n - 1; j >= 0; j--){
                int left = Integer.MAX_VALUE, down = Integer.MAX_VALUE, right = Integer.MAX_VALUE;
                if (j < n -1) right = matrix[i + 1][j + 1];
                if (j > 0) left = matrix[i + 1][j - 1];
                down = matrix[i + 1][j];
                matrix[i][j] += Math.min(left, Math.min(down, right));
            }
        }
        int ans = Integer.MAX_VALUE;
        for (int i = 0; i < n; i++){
            ans = Math.min(ans, matrix[0][i]);
        }
        return ans;
    }
}